### PR TITLE
Use Setuptools

### PR DESCRIPTION
**Pull request by SchnWalter**
Related to #77

### CODE REVIEW

File: main/terra [hunk 1]
```diff
@@ -18,15 +18,12 @@
...
-except:
-    print "Error Terra library is not installed."
-    sys.exit(0)
+# Add the project root the the PYTHONPATH environment variable.
+ROOT = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
+sys.path.append(ROOT)
 
-terra_main.main()
+if __name__ == '__main__':
+    # TODO: Update setup.py to also set LIBOVERLAY_SCROLLBAR.
```
Comment (Sixdsn) [inline main/terra]: This has to be done before merging

File: .gitignore [hunk 2]
```diff
@@ -1,26 +1,18 @@
...
+# Ignore the workspace files.
+/po/
+*--ignore
+*--ignore.*
+*gz
+
+# Ignore distribution and packaging files.
+/build/
+/dist/
+/*.egg-info/
```
Comment (Sixdsn) [inline .gitignore:18]: I think we can add too:
/debian/files
/debian/terra*
/lextab.py

File: setup.py [hunk 3]
```diff
@@ -118,49 +61,70 @@ def parse_pkg_config(command, components, options_dict=None):
...
-     ),
-        ('bin/',
-         glob.glob("main/terra")
-     ),
+        ('share/icons/hicolor/scalable/apps', ['terra/resources/terra.svg']),
+        ('share/applications', ['terra/resources/terra.desktop'])
+    ],
+    install_requires=[
+        'gi',
+    ],
```
Comment (Sixdsn) [inline setup.py]: We need to remove this

File: terra/VteObject.py [hunk 4]
```diff
@@ -55,15 +54,22 @@
     "(?:news:|man:|info:)[[:alnum:]\\Q^_{|}~!\"#$%&'()*+,./;:=?`\\E]+"]
 
 class VteObjectContainer(Gtk.HBox):
-    def __init__(self, parent, bare=False, progname=ConfigManager.get_conf('general', 'start_shell_program'), pwd=None):
+    def __init__(self, parent, bare=False, progname=None, pwd=None):
```
Comment (Sixdsn) [inline terra/VteObject.py:57]: Why do you prefer to remove the default parameter and check if progname is set later?
Is it about readability?

File: terra/config.py [hunk 5]
```diff
@@ -18,39 +18,50 @@
...
+        if os.path.exists(config_path):
+            config_file_path = os.path.join(config_path, cls.config_file_name)
+            if os.path.exists(config_file_path):
+                cls.config.read(config_file_path)
 
-    @staticmethod
-    def init():
-        ConfigParser.SafeConfigParser({})
-        ConfigManager.save_config()
+        # TODO: Remove once the code no longer requires the config files!
```
Comment (Sixdsn) [inline terra/config.py:53]: I don't figure out what you mean?

File: terra/config.py [hunk 6]
```diff
@@ -18,39 +18,50 @@
...
+        cls.save_config()
 
-    @staticmethod
-    def get_sections():
-        return (ConfigManager.config.sections())
+    @classmethod
+    def get_sections(cls):
+        # TODO: Remove! For some reason, on my setup the `sections()` method
+        # raises a "tuple assignment index out of range" exception.
+        # Reading items from the config seems to solve the issue.
```
Comment (Sixdsn) [inline terra/config.py:61]: Yep i found this pretty weird when I did it but I couldn't find a better solution...

File: terra/preferences.py [hunk 7]
```diff
@@ -18,21 +18,32 @@
...
+
+from terra.config import ConfigManager
+from terra.handler import TerraHandler
+from terra.i18n import t
 
-class Preferences():
 
+class Preferences:
     def __init__(self):
+        # TODO: Only initialize the UI once. Find a way to hide the window.
```
Comment (Sixdsn) [inline terra/preferences.py:34]: Issue #82

**Comment by Sixdsn** [comment]
This seems great! :)
I prefer to merge it when all the TODO stuff is done, except for the last one which was existing so i opened a new issue.
Feel free to answer to my notes ;)

File: .gitignore (hunk 2 above)
Comment (Sixdsn) [inline .gitignore:18]: Done

File: setup.py [hunk 8]
```diff
@@ -118,49 +61,67 @@ def parse_pkg_config(command, components, options_dict=None):
...
-        ('share/terra/ui/',
-         glob.glob("data/ui/*.ui")
-     ),
-        ('bin/',
-         glob.glob("main/terra")
-     ),
+        ('share/icons/hicolor/scalable/apps', ['terra/resources/terra.svg']),
+        ('share/applications', ['terra/resources/terra.desktop'])
+    ],
+    classifiers=[
```
Comment (Sixdsn) [inline setup.py:115]: I removed the **gi** requirement

File: terra/VteObject.py (hunk 4 above)
Comment (SchnWalter) [inline terra/VteObject.py:57]: Yes, for readability mainly.  But I want to change the code so that you don't need so many dependencies upon `from terra.xyz imports ClassXYZ`.

File: terra/config.py (hunk 5 above)
Comment (SchnWalter) [inline terra/config.py:53]: There should be no need for a user config file (`~/.config/terra/main.cfg`) when you just initialize the ConfigManager, because you don't make any changes, the application should run just fine without that file. It should be created only when you change the settings.

File: setup.py (hunk 8 above)
Comment (SchnWalter) [inline setup.py:115]: :+1:

File: terra/config.py (hunk 6 above)
Comment (SchnWalter) [inline terra/config.py:61]: I solved it on the new manager I created. Maybe I'll just commit it as `settings.py`.  The new one seems to work just fine, 

I replace the whole thing with this:

```
    @property
    def sections(self):
        """
        :return dict:
        """
        sections_list = self._defaults.keys()
        for section in self.iterkeys():
            if section not in sections_list:
                # Delete default values.
                if sections_list[section] is None:
                    del sections_list[section]
                    continue

                sections_list.append(section)

        return sections_list
```

Where that "Delete default values" part simply allows one to remove a whole section from the default settings, for example a layouts section. But it should not really be required to have such a thing, maybe I'll just remove it.

File: terra/config.py (hunk 5 above)
Comment (Sixdsn) [inline terra/config.py:53]: Well everything that is stored in the ConfigManager is used for layout and preferences (coming from the `main.cfg` file) so it's needed on startup right?

File: terra/config.py (hunk 6 above)
Comment (Sixdsn) [inline terra/config.py:61]: Yes it's usefull when the user saves it's layout after closing a window so the section is removed from the configuration and won't be created at next launch.
Open a new issue for this module ;)

File: terra/VteObject.py (hunk 4 above)
Comment (Sixdsn) [inline terra/VteObject.py:57]: Ok
thoses important changes will be in another pull request then ;)

**Comment by Sixdsn** [comment]
Now it seems quite ready to merge.
For the terra/config.py we will open an issue to enhance that.
Do you think it's okay to merge now?

File: terra/config.py (hunk 5 above)
Comment (SchnWalter) [inline terra/config.py:53]: Everything needed is already inside `ConfigDefaults`, so if there is no user settings file, you don't need to create one, until you need to save some new settings.

**Comment by SchnWalter** [comment]
All good on this PR :+1:
